doc/source/development/rfc/rfc104_gdal_cli.rst

```diff
@@ -0,0 +1,1175 @@
+      --bbox <BBOX>                                        Bounding box as xmin,ymin,xmax,ymax [may be repeated]
+
+    * reproject [OPTIONS]
+    ---------------------
+
+    Reproject.
+
+    Options:
+      -s, --src-crs <SRC-CRS>                              Source CRS
+      -d, --dst-crs <DST-CRS>                              Destination CRS [required]
+
+    * write [OPTIONS] <OUTPUT>
+    --------------------------
+
+    Write a vector dataset.
+
+    Positional arguments:
+      -o, --output <OUTPUT>                                Output vector dataset [required]
+
+    Options:
+      -f, --of, --format, --output-format <OUTPUT-FORMAT>  Output format
+      --co, --creation-option <KEY=VALUE>                  Creation option [may be repeated]
+      --lco, --layer-creation-option <KEY=VALUE>           Layer creation option [may be repeated]
+      --overwrite                                          Whether overwriting existing output is allowed
+      --update                                             Whether updating existing dataset is allowed
+      --overwrite-layer                                    Whether overwriting existing layer is allowed
+      --append                                             Whether appending to existing layer is allowed
+      -l, --output-layer <OUTPUT-LAYER>                    Output layer name
+
+
+CLI specification
+-----------------
+
+Subcommand syntax
++++++++++++++++++
+
+.. code-block:: shell
+
+        gdal <subcommand> [<subsubcommand>]... [<options>]... [<positional arguments>]...
+
+where subcommand is something like ``raster``, ``vector``, etc. with potential
+sub-subcommand like ``info``, ``convert``, etc.
+
+Option naming conventions
++++++++++++++++++++++++++
+
+* One-letter short names preceded with dash: ``-z``
+
+  When a value is specified, it must be separated with a space: ``-z <value>``
+
+* Longer names preceded with two dashes and using dash to separate words,
+  lower-case capitalized: ``--long-name``
+
+  When a single value is expected, it must be separated with a space or equal sign:
+  ``--long-name <value>`` or ``--long-name=<value>``.
+
+  In the rest of the document, we will use the version with a space separator,
+  but equal sign is also accepted.
+
+Repeated values / multi-valued options
+++++++++++++++++++++++++++++++++++++++
+
+Existing GDAL command line utilities have an inconsistent strategy regarding
+how to specify repeated values (band indices, nodata values, etc.), sometimes
+with the switch being repeated many times, sometimes with a single switch but
+the values being grouped together and separated with spaces or commas.
+
+With this RFC, when several values are expected (or allowed), they must be
+separated with a ``,`` (comma): ``--bbox <xmin>,<ymin>,<xmax>,<ymax>``
+
+An alternative when passing several values is also to repeat the option as
+many times as needed: ``--co KEY1=VALUE1 --co KEY2=VALUE2`` and
+``--co KEY1=VALUE1,KEY2=VALUE2``.
```

Comment:
This examples are confusing to me.
`--bbox` takes always 4 and only 4 numbers (separated by commas). It is not an example of repeated option like `--co`. 

Related to `--co`, again for simplicity and better understanding, I would like only one way to do it: repeating `--co` several times. That leaves the comma for cases like `--bbox`, where the parameters are related (the 4-tuple w,s,e,n)

Reply:
> `--bbox` takes always 4 and only 4 numbers (separated by commas). It is not an example of repeated option like `--co`.

They are handled the same way by the GDALAlgorithm argument parser.  --bbox has mincount=maxcount=4 and --co has mincount=0 and maxcount=unlimited.

Reply:
>  So that nobody trips over usage like --bbox <ymax> --bbox <xmax> --bbox <xmin> --bbox <ymin>

The advantage of having my "own" parser is that I can indeed add a mode to disallow that. Will do